python/Ganga/GPIDev/Lib/File/LocalFile.py

```diff
@@ -224,6 +224,19 @@ def remove(self):
 
         return
 
+    def get(self, localPath=''):
```

Comment:
I think this highlights the difficulty of working with `IGangaFile` given its dual role as both input and output files which is especially compounded by the special case of `LocalFile`.

For a standard `IGangaFile`, `put` should copy the file from the local machine to some remote place and `get` should copy the file from come remote place to the local machine. Normally a job using a `IGangaFile` as an input would read the file *from* the remote location whereas a job using it as an output would write it *to* there when done. When used as an inputfile the schema attributes need to be interpreted differently to when it's an output file: e.g. `MassStorageFile`'s `namePattern` is used to find the file on the remote file system when it's an input file but is used to find the file on the WN system when it's an outputfile.

As well as input/output differences there's also the difference between file handling done on the WN (e.g. `DiracFile`) and files sent back via the sandbox to be handled on the client (e.g. `LocalFile`). There's different codepaths and potentially different meanings to the schema attributes.

Add to this the complications of subfiles and filename globbing the whole system is very unintuitive. We need to be very clear on the API to which we expect `IGangaFile`s to conform and exactly how it should be used before we start implementing it. For example, we already have a way for the user to specify where the file should be copied to in the `LocalFile`/output situation which is `localDir`. Adding another argument to `get()` just confuses this. What should be done instead is make sure that all `IGangaFile` subclasses have a `localDir` schema attribute and make sure that is used consistently; It's always the source directory for a `put()` and the destination directory for a `get()`. Likewise, for `namePattern` (only ever a file glob within `localDir` for inputfile/`put`, within some WN dir for an outputfile and within the remote dir for a `get`).